### PR TITLE
Add async support

### LINKED ISSUE #80: Is there a asynchronous version of solana-py?
Is there a asynchronous version of solana-py?

### PR DESCRIPTION
**Pull request by kevinheavey**
Closes #80 

Summary of changes:

- Adds `AsyncClient` and `AsyncToken` classes.
- Refactors sync code so the async code can use it with minimal duplication. No changes to logic or behaviour of sync methods.
- Adds tests for async methods.
- Adds async example to the README
- Adds httpx as a dependency.
- Fixes `test_get_identity` which I noticed wasn't testing `get_identity`. Instead it was redundantly testing `get_genesis_hash`.
- Bumps minor version number.

### CODE REVIEW

**Comment by kevinheavey** [comment]
Should be good to run the CI again. By the way, any idea why those two docstrings didn't show up when I ran `make lint`? I'm pretty sure that command caught similar errors for me before.

File: solana/rpc/api.py [hunk 1]
```diff
@@ -67,13 +401,1039 @@ def get_balance(self, pubkey: Union[PublicKey, str], commitment: Optional[Commit
...
+        if not resp["result"]:
+            raise Exception("Unable to confirm transaction %s" % tx_sig)
+        err = resp.get("error") or resp["result"].get("meta").get("err")
+        if err:
+            self._provider.logger.error("Transaction error: %s", err)
+
+        return resp
+
+
+class AsyncClient(_ClientCore):  # pylint: disable=too-many-public-methods
```
Comment (michaelhly) [inline solana/rpc/api.py:1390]: Would it be possible to split into three different files?

- solana/rpc/core.py
- solana/rpc/api.cpy (old synchornous client)
- solana/rpc/async_api.py (new async client)

File: solana/rpc/providers/http.py [hunk 2]
```diff
@@ -2,60 +2,123 @@
...
             response.raise_for_status()
         except (IOError, requests.HTTPError) as err:
             self.logger.error("Health check failed with error: %s", str(err))
             return False
 
         return response.ok
+
+
+class AsyncHTTPProvider(AsyncBaseProvider, _HTTPProviderCore):
+    """Async HTTP provider to interact with the http rpc endpoint."""
```
Comment (michaelhly) [inline solana/rpc/providers/http.py:85]: Same here?

File: tests/integration/test_http_client.py [hunk 3]
```diff
@@ -2,18 +2,25 @@
...
+from solana.rpc.api import DataSliceOpt, AsyncClient
 from solana.transaction import Transaction
 
-from .utils import assert_valid_response, confirm_transaction
+from .utils import assert_valid_response, confirm_transaction, aconfirm_transaction, compare_responses_without_ids
 
 
 @pytest.mark.integration
-def test_request_air_drop(stubbed_sender, test_http_client):
+def test_request_air_drop(stubbed_sender, test_http_clients):
```
Comment (michaelhly) [inline tests/integration/test_http_client.py:12]: Can we test the async client separate from the old client? Having both synchronous and async tests in the same method makes maintenance more difficult.

**Review by michaelhly — COMMENTED** [review]
Hey @kevinheavey. Thank you for contributing. This PR looks great. 

I think it would be better if we have clearer boundaries between the old synchronous client and the new async client so that this codebase is easier to maintain. 

And that if someone decides implement a websocket client, they should not be cramming into the existing `api.py` file nor the existing integration tests.

File: solana/rpc/api.py (hunk 1 above)
Comment (kevinheavey) [reply solana/rpc/api.py:1390]: Will do

File: tests/integration/test_http_client.py (hunk 3 above)
Comment (kevinheavey) [reply tests/integration/test_http_client.py:12]: This is possible, but we'll have a lot of duplicated code without a big refactor of the test suite. Plus there are places where we just test that the response the async client received is the same as that of the sync client, and we'll lose that (without more refactoring).

To get this over the line I'll just duplicate all the old pre-async tests and make async versions of them. Your call if that's what you want

File: solana/rpc/api.py (hunk 1 above)
Comment (kevinheavey) [reply solana/rpc/api.py:1390]: In the same spirit, I'll give AsyncToken its own file too

File: tests/integration/test_http_client.py (hunk 3 above)
Comment (michaelhly) [reply tests/integration/test_http_client.py:12]: Yes. Duplicate tests are fine, it's just less confusing. We can decide to refactor and try to reduce duplicates later.

**Comment by kevinheavey** [comment]
Made those changes.
Extras:
- Removed pytest-tornasync as a dev dependency. It wasn't doing anything until I added async tests using pytest-asyncio, at which point it started confusing pytest.
- Fixed typo in tests: reciever -> receiver

**Comment by kevinheavey** [comment]
Forgot the linters, should work now

**Comment by kevinheavey** [comment]
Think pylint should be okay now

**Comment by michaelhly** [comment]
Let's just disable `R0913: Too many arguments (7/5) (too-many-arguments)` for those modules. 

https://pylint.readthedocs.io/en/latest/faq.html#is-there-a-way-to-disable-a-message-for-a-particular-module-only

**Comment by kevinheavey** [comment]
Done. Got a good feeling about this one. Pylint is not complaining about anything I touched

**Review by michaelhly — APPROVED** [review]
Thanks, @kevinheavey. This is great!

**Pull request merged**